flocker/testtools/__init__.py

```diff
@@ -299,6 +302,28 @@ def loop(failure):
     return d
 
 
+def retry_effect_with_timeout(effect, timeout):
+    """
+    If ``effect`` fails, retry it until ``timeout`` expires.
+
+    :param Effect effect: The Effect to retry.
+    :param int timeout: Timeout in seconds.
+    """
+
+    end_time = time.time() + timeout
+
+    def ret(e):
+        if time.time() > end_time:
+            return Effect(Constant(False))
+        else:
+            time.sleep(ret.wait_secs)
```

Comment:
I'm not super familiar with effect, but I think that rather than explicitly sleeping here, we should be using `effect.Delay` (see http://effect.readthedocs.org/en/latest/apidocs.html?highlight=time#effect.Delay) and presumably sequencing it somehow (I don't know and the docs don't say, but I'd start by looking for other examples. If you'd like we can pair on this).

Once you've done _that_, then it becomes feasible to write a few quick unit tests for this, which you really should.